### PR TITLE
HoareOptimizer pass does preserve state between runs anymore

### PR DESCRIPTION
**Pull request by pierrepocreau**
### Summary

Changes on `HoareOptimizer` to solve #4999 

### Details and comments

Reset `HoareOptimizer` internal state after each run, so that it can run multiple times, without having to create a fresh instance each times.

### CODE REVIEW

File: qiskit/transpiler/passes/optimization/hoare_opt.py [hunk 1]
```diff
@@ -338,4 +348,5 @@ def run(self, dag):
         if self.size > 1:
             for qbt in dag.qubits:
                 self._multigate_opt(dag, qbt.index)
+        self._reset()
```
Comment (1ucian0) [inline qiskit/transpiler/passes/optimization/hoare_opt.py:351]: Shouldn't `self._reset` be before `self._initialize`. I'm surprised this works... maybe I'm missing something.

**Review by 1ucian0 — COMMENTED** [review]
Thanks for having a look to this!

What do you think that, instead of resetting, making `__init__` "lighter". For example, `__init__` only set the size and move all the other attributes as methods in `run`. For example, for `self.solver`:
```python
    def __init__(self, size=10):
        super().__init__()
        self.size = size

    def run(self, dag):
        solver = Solver()
        self._initialize(dag, solver)
        self._traverse_dag(dag, solver)
        ...
```

I think this might be less error prone, but I don't know if it will clutter the method calls with tons of params. What do you think?

File: qiskit/transpiler/passes/optimization/hoare_opt.py (hunk 1 above)
Comment (pierrepocreau) [reply qiskit/transpiler/passes/optimization/hoare_opt.py:351]: You are right, `self._reset` would be better placed before `self._initialize`. Right now it works but we clean the internal state after the run, while we might want to keep the internal state, and clean it only if we call another run.

**Comment by pierrepocreau** [comment]
So I thought about it, and honestly I think it is better to let the attributes in the `__init__`, for I think the methods accessing all the dicts by `self. ...` is easy to track and understand. Passing the dicts as methods arguments, we would have intertwined methods mutating dicts passed as parameters, while some of the methods are also returning values, it would probably be hard to understand.

**Comment by 1ucian0** [comment]
Maybe now `__init__` can initiate the attributes to `None`?

```python
    def __init__(self, size=10):
        super().__init__()
        self.solver = Solver()
        self.variables = dict()
        self.gatenum = dict()
        self.gatecache = dict()
        self.varnum = dict()
        self.size = size
```

**Review by 1ucian0 — APPROVED** [review]
Thanks a lot for the fix and welcome to the community!

**Pull request merged**